dockers/texcmp/texcmp.js

```diff
@@ -0,0 +1,244 @@
+    // dirs have been created, template has been read, now rasterize.
+    return Q.all(todo.map(processTestCase));
+}).done();
+
+// Process a single test case: rasterize, then create diff
+function processTestCase(key) {
+    if (blacklist[key]) return;
+    var url = data[key];
+    var query = url.replace(/^.*?\?/, ""); // extract query string
+    query = query.replace(/\+/g, "%2B"); // plus doesn't mean space here
+    query = querystring.parse(query);
+    var tex = "$" + query.m + "$";
+    if (query.display) {
+        tex = "$$" + query.m + "$$";
+    }
+    if (query.pre) {
+        tex = query.pre.replace("<br>", "\\\\") + tex;
+    }
+    if (query.post) {
+        tex = tex + query.post.replace("<br>", "\\\\");
+    }
+    tex = template.replace(/\$.*\$/, tex.replace(/\$/g, "$$$$"));
+    var texFile = path.join(tmpDir, key + ".tex");
+    var pdfFile = path.join(tmpDir, key + ".pdf");
+    var pngFile = path.join(teximgDir, key + "-pdflatex.png");
+    var browserFile = path.join(imagesDir, key + "-firefox.png");
+    var diffFile = path.join(diffDir, key + ".png");
+
+    // Step 1: write key.tex file
+    var fftLatex = Q.nfcall(fs.writeFile, texFile, tex).then(function() {
+        // Step 2: call "pdflatex key" to create key.pdf
+        return execFile("pdflatex", [
+            "-interaction", "nonstopmode", key
+        ], {cwd: tmpDir});
+    }).then(function() {
+        console.log("Typeset " + key);
+        // Step 3: call "convert ... key.pdf key.png" to create key.png
+        return execFile("convert", [
+            "-density", dpi, "-units", "PixelsPerInch", "-flatten",
+            pdfFile, pngFile
+        ]);
+    }).then(function() {
+        console.log("Rasterized " + key);
+        // Step 4: apply FFT to that
+        return fftPNG(pngFile);
+    });
+    // Step 5: apply FFT to reference image as well
+    var fftBrowser = fftPNG(browserFile);
+
+    return Q.all([fftBrowser, fftLatex]).spread(function(browser, latex) {
+        // Now we have the FFT result from both 
+        // Step 6: find alignment which maximizes overlap.
+        // This uses a FFT-based correlation comoputation.
+        var x, y;
+        var real = newMatrix();
+        var imag = newMatrix();
+
+        // Step 6a: (real + i*imag) = latex * conjugate(browser)
+        for (y = 0; y < alignHeight; ++y) {
+            for (x = 0; x < alignWidth; ++x) {
+                var br = browser.real.get(y, x);
+                var bi = browser.imag.get(y, x);
+                var lr = latex.real.get(y, x);
+                var li = latex.imag.get(y, x);
+                real.set(y, x, br * lr + bi * li);
+                imag.set(y, x, br * li - bi * lr);
+            }
+        }
+
+        // Step 6b: (real + i*imag) = inverseFFT(real + i*imag)
+        fft(-1, real, imag);
+
+        // Step 6c: find position where the (squared) absolute value is maximal
+        var ox = 0;
+        var oy = 0;
+        var ov = -1;
```

Comment:
`ox` and `oy` stand for offset, yes. `ov` is the value associated with the best offset, or to be more precise, the squared absolute value. Since that will never be negative, initializing it to a negative number ensures that the first actual pixel will overwrite that value.